### ISSUE TITLE
Custom Resolution and Filter Bugs

### ISSUE THREAD

**Issue by amadeus** [photonstorm/phaser#2320]
Hello, I chatted a bit with @photonstorm over slack about these two specific issues and am following up with an issue to easily recreate them.

I have made a [demo page](http://wtf.amadeus.wtf/phaser-filter-bug-demo/#4) that can be used to see these two bugs in action. This demo page has some interactive dropdowns to change the in game resolution and toggle various settings that I will refer to later in this issue. I have also created a gist of my [state file](https://gist.github.com/amadeus/b9e0ed4ccf23dd88917b) that can be easily referred too.

Also, I believe both of these issues are related to the version of PIXI you manage internally for Phaser 2.x.

## Bug 1:

Filters are rendered differently depending on whether the filter is the last in the filters array or not.

If you check the [demo page](http://wtf.amadeus.wtf/phaser-filter-bug-demo/#4) and look closely, I have am applying a [noise filter to a background image](https://gist.github.com/amadeus/b9e0ed4ccf23dd88917b#file-menu-js-L89). This scene (using the link above) that is being rendered at a resolution of 4. The noise effect however looks a bit blurry (you may have to zoom in or screenshot and zoom in to see the blurriness). If you click the `Add Passthrough` button, I [add another filter](https://gist.github.com/amadeus/b9e0ed4ccf23dd88917b#file-menu-js-L58) to the background image and you will see that now the noise is nice and sharp and pixellated. Just for note, this second filter I apply actually does nothing, it just returns the colors of the scene exactly as they are ([passthrough filter](https://gist.github.com/amadeus/00e3da591a3bbf32d11e)).

It appears as if the all filters before the last take resolution into account properly and don't scale up. I've tested this with a variety of different filters all applied at the same time, and the only blurry one is ALWAYS the last one. I believe it has something to do with the [pushFilter method](https://github.com/photonstorm/phaser/blob/db641ca82e608470bd7902de3447d048d9715ab5/src/pixi/renderers/webgl/utils/WebGLFilterManager.js#L152) since a different path is taken for all but the [last filter](https://github.com/photonstorm/phaser/blob/db641ca82e608470bd7902de3447d048d9715ab5/src/pixi/renderers/webgl/utils/WebGLFilterManager.js#L192)

Here is a blown up image of the noise with no passthrough (as the page loads):
![no passthrough](https://dl.dropboxusercontent.com/s/tedwjle8bvfpmzp/no-passthrough.png?dl=0)

And here is a blown up image of the noise with the pass through filter appended to the filters array (after pressing the `Add Passthrough` button):
![with passthrough](https://dl.dropboxusercontent.com/s/9kztdnxzo285wf7/passthrough.png?dl=0)

Bear in mind, this issue ONLY exists if the resolution is greater than 1. Of you set the resolution to 1 nothing changes based on whether the passthrough filter has been added or  not.

## Bug 2:

By [default](https://gist.github.com/amadeus/b9e0ed4ccf23dd88917b#file-menu-js-L75), the scene has 2 images added to a group. One of those images has a noise filter.

If you now add a filter to the entire stage, using the `Add Scene Filter` button ([code here](https://gist.github.com/amadeus/b9e0ed4ccf23dd88917b#file-menu-js-L63)), AND there is a custom resolution greater than 1, every child element gets scaled up relative to the resolution. You can test this by trying out different resolution combinations coupled with the `Add/Remove Scene Filter` button.

And for reference, here is what the scene looks like at a resolution of 2, with no stage filter (the correct view):
![no stage filter](https://dl.dropboxusercontent.com/s/4unvsy42ss62hbf/no-stage-filter.png?dl=0)

And here's the same scene, with a filter added to the stage at a resolution of 2:
![with stage filter](https://dl.dropboxusercontent.com/s/8r79pr5qsrj9upe/stage-filter.png?dl=0)

And finally, here is what it SHOULD look like, with the stage filter, however this is only possible at a resolution of 1:
![with stage filter, at 1x](https://dl.dropboxusercontent.com/s/snxm1yxhl9zd525/stage-filter-1x.png?dl=0)

I've also done testing with putting the filters on the group instead of the stage, and the same results occur. I think the bug occurs when I have filters on nested elements. If I add filters to sibling elements, nothing seems to get messed up.

**Comment by englercj** [comment]
Ok, so I'm looking into this now. Here is some information:

**Bug 1:**

The reason you see it become hard-edged is because when you have only 1 single filter it draws through the normal pipeline. That is, it just applies the filter to the backbuffer. However, when you have multiple filter passes (even just a passthrough) it has to render to an FBO first and then use that FBO as input to the next filter. Rending in FBO space instead of backbuffer space means that you lose anti-aliasing, OpenGL 2.0 doesn't support MSAA filtering on FBOs without a special extension. That extension is core in WebGL 2.0, but for now it in an extension (GL_EXT_framebuffer_multisample) and I'm not sure how widely supported the extension is. We usually avoid them because support is often pretty low.

**Bug 2:**

See #2326 :)

**Comment by amadeus** [comment]
@englercj Dude you rock!

In regards to bug 1: I have `antialias:false`, so I actually prefer the look WITH the pass through filter. Shouldn't it be respecting that property on that last filter?

**Comment by englercj** [comment]
@amadeus Ok, I updated my PR to fix one more place that I had missed that wasn't taking resolution into account. That solves Bug 2 I think.

As for Bug 1, it was being caused by the same problem I fixed in this last commit. Basically the output texture wasn't being sized properly so it was scaling it up after running the filter meaning you got the pixelation.

After this fix, if you want your noise to be more pixelated just modify the filter to be less granular. It seems to be working fine for me now but I easily could have missed something
